sql/core/src/main/scala/org/apache/spark/sql/execution/Exchange.scala

```diff
@@ -208,6 +208,19 @@ private[sql] case class EnsureRequirements(sqlContext: SQLContext) extends Rule[
     }
   }
 
+  private def removeRedundantShuffle(childPlan: SparkPlan): SparkPlan = {
+    // For an query like df1.repartition(1000).join(df2, "_1"), we will repartition df1
+    // with shuffle. Then we perform shuffle again as part of the exchange added here.
+    // To avoid the extra shuffle, we should remove repartition operators if they are
+    // the child of Exchange and shuffle=True.
+    val optimized = childPlan match {
+      case Repartition(n, true, plan) => plan
+      case Exchange(p, plan) => plan
```

Comment:
How can this case ever occur, given that Exchange operators are currently added all at once in a single pass? Are you planning ahead in case we implement Repartition using Exchange, as in #8030?